pkg/hub/controllers/vpnkeyrotation/reconciler.go

```diff
@@ -0,0 +1,394 @@
+			vpnKeyRotation.Status.StatusHistory = make(map[string][]hubv1alpha1.StatusOfKeyRotation)
+		}
+		history := vpnKeyRotation.Status.StatusHistory[gatewayName]
+
+		// Prepend current rotation to history, maintaining a maximum of 5 rotations
+		// StatusHistory is a stack n number of gateways rotation status
+		history = append([]hubv1alpha1.StatusOfKeyRotation{rotation}, history...)
+		if len(history) >= 5 {
+			// Remove the oldest rotation from the history
+			history = history[:5]
+		}
+		vpnKeyRotation.Status.StatusHistory[gatewayName] = history
+		return r.Status().Update(ctx, vpnKeyRotation)
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) syncCurrentRotationState(ctx context.Context,
+	vpnKeyRotation *hubv1alpha1.VpnKeyRotation, allGwsUnderCluster []string) error {
+	log := logger.FromContext(ctx)
+	// cluster deregister handling
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		if getErr := r.Get(ctx,
+			types.NamespacedName{Name: vpnKeyRotation.Name, Namespace: vpnKeyRotation.Namespace},
+			vpnKeyRotation); getErr != nil {
+			return getErr
+		}
+		syncedRotationState := make(map[string]hubv1alpha1.StatusOfKeyRotation)
+		for _, elem := range allGwsUnderCluster {
+			if obj, ok := vpnKeyRotation.Status.CurrentRotationState[elem]; ok {
+				syncedRotationState[elem] = obj
+			}
+		}
+		if len(syncedRotationState) != len(vpnKeyRotation.Status.CurrentRotationState) {
+			log.V(3).Info("syncing current rotation state for the gateways")
+			vpnKeyRotation.Status.CurrentRotationState = syncedRotationState
+			return r.Status().Update(ctx, vpnKeyRotation)
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *Reconciler) updateInitialRotationStatusForAllGws(ctx context.Context, vpnKeyRotation *hubv1alpha1.VpnKeyRotation,
+	allGwsUnderCluster []string) error {
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		if getErr := r.Get(ctx,
+			types.NamespacedName{Name: vpnKeyRotation.Name, Namespace: vpnKeyRotation.Namespace},
+			vpnKeyRotation); getErr != nil {
+			return getErr
+		}
+		m := make(map[string]hubv1alpha1.StatusOfKeyRotation)
+		for _, selectedGw := range allGwsUnderCluster {
+			m[selectedGw] = hubv1alpha1.StatusOfKeyRotation{
+				Status:               hubv1alpha1.Complete,
+				LastUpdatedTimestamp: vpnKeyRotation.Spec.CertificateCreationTime,
+			}
+		}
+		vpnKeyRotation.Status.CurrentRotationState = m
+		return r.Status().Update(ctx, vpnKeyRotation)
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func isServer(sliceGw *kubeslicev1beta1.SliceGateway) bool {
+	return sliceGw.Status.Config.SliceGatewayHostType == "Server"
+}
+
+func (r *Reconciler) updateCertificates(ctx context.Context, rotationVersion int, sliceGw string,
+	req reconcile.Request) (ctrl.Result, bool, error) {
+	log := logger.FromContext(ctx)
+	currentSecretName := sliceGw + "-" + strconv.Itoa(rotationVersion)
```

Comment:
current limit is 253 chars which is significant according to me. So i think we should not worry about this now
If we face this issue we can revisit it. But i think we are putting some limit at controller side itself while generating this secret. so this will not harm anything